sec_certs/model/dependency_vulnerability_finder.py

```diff
@@ -0,0 +1,109 @@
+
+
+class DependencyVulnerabilityFinder:
+    def __init__(self, certificates: Certificates):
+        self.vulnerabilities: Vulnerabilities = {}
+        self.certificates = certificates
+
+    def _get_dataset_cert_ids_occurrences(self) -> Dict[str, int]:
+        cert_id_occurrences: Dict[str, int] = {}
+
+        for dgst in self.certificates:
+            cert_id = self.certificates[dgst].heuristics.cert_id
+
+            if cert_id is None:
+                continue
+
+            cert_id_occurrences[cert_id] = cert_id_occurrences.get(cert_id, 0) + 1
+
+        return cert_id_occurrences
+
+    def _get_cert_dependency_cves(self, dgst: str, dependency_type: DependencyType) -> Optional[Set[str]]:
+        dependency_type_dict = {
+            DependencyType.DIRECT: self.certificates[dgst].heuristics.report_references.directly_referenced_by,
+            DependencyType.INDIRECT: self.certificates[dgst].heuristics.report_references.indirectly_referenced_by,
+        }
+
+        references = dependency_type_dict[dependency_type]
+
+        if not references:
+            return None
+
+        vulnerabilities = set()
+        dataset_cert_id_occurrences = self._get_dataset_cert_ids_occurrences()
+
+        for cert_id in references:
+            if cert_id is None:
+                continue
+
+            cert_id_occurrences = dataset_cert_id_occurrences.get(cert_id)
+
+            if cert_id_occurrences is None or cert_id_occurrences >= 2:
+                continue
+
+            for dgst in self.certificates:
+                cert_obj = self.certificates[dgst]
+                cert_obj_cves = cert_obj.heuristics.related_cves
+
+                if cert_obj.heuristics.cert_id == cert_id and cert_obj_cves:
+                    vulnerabilities.update(cert_obj_cves)
+
+        return vulnerabilities if vulnerabilities else None
+
+    def fit(self) -> Vulnerabilities:
+        cert_id_occurrences = self._get_dataset_cert_ids_occurrences()
+        thrown_away_cert_counter = 0
+
+        for dgst in self.certificates:
+            cert_id = self.certificates[dgst].heuristics.cert_id
+
+            if cert_id is None:
+                continue
+
+            if cert_id_occurrences[cert_id] >= 2:
+                thrown_away_cert_counter += 1
+                continue
+
+            self.vulnerabilities[dgst] = {}
+            self.vulnerabilities[dgst][DependencyType.DIRECT.value] = self._get_cert_dependency_cves(
+                dgst, DependencyType.DIRECT
+            )
+            self.vulnerabilities[dgst][DependencyType.INDIRECT.value] = self._get_cert_dependency_cves(
+                dgst, DependencyType.INDIRECT
+            )
+
+        if thrown_away_cert_counter > 0:
+            logging.warning("There were total of %s certificates skipped due to duplicity", thrown_away_cert_counter)
+
+        return self.vulnerabilities
+
+    def get_dependency_vulnerabilities(self, dgst: str) -> DependencyCVE:
```

Comment:
**Suggestion**: Rename to `predict_sigle_cert()`

we should also add a method `predict()` that would, given a list of certificates (or dgsts, idk), predict all the vulnerabilities. Again, this is to adhere with how sklearn classifiers typically work.

Reply:
It would also be great if you could rename/create these methods for the reference finder that you prevously implemented. I missed it that time...